### ISSUE TITLE
Recursive future stack-overflows

### ISSUE THREAD

**Issue by mxcl** [apple/swift-nio#970]
I have a long, long, long chain of futures, when finally I resolve it, it stack-overflows.

If I reduce the chain enough, the app does not crash.

I understand that maybe I shouldn't be making a vast chain (probably 1,000 thens) however, the API allows it, so I opened this ticket in case you agree and think it should be mitigated.

I attach the top and bottom of the stack, the crash happens in different functions, which I am assuming because it varies how large the stack needs to be before it is too much for whatever system cannot handle the size of the call stack.

Also, maybe you have a suggestion for how I can work around this for now?

<img width="265" alt="Screen Shot 2019-04-12 at 7 22 15 PM" src="https://user-images.githubusercontent.com/58962/56070869-7f8d9080-5d58-11e9-8ed3-45e4ed70ce98.png">
<img width="267" alt="Screen Shot 2019-04-12 at 7 22 22 PM" src="https://user-images.githubusercontent.com/58962/56070872-80bebd80-5d58-11e9-9d9c-0d79c96f4d29.png">

**Comment by mxcl** [comment]
The top 95% repeats between `_run()` and `_setValue`

**Comment by mxcl** [comment]
Here's an example of the repeating portion:

<img width="262" alt="Screen Shot 2019-04-12 at 7 33 36 PM" src="https://user-images.githubusercontent.com/58962/56071074-e95a6a00-5d59-11e9-8803-119fa51ecc8b.png">

**Comment by Lukasa** [comment]
Yup, this is an issue we knew we were going to have to tackle eventually. The API does allow this, and we definitely should mitigate it. 

In the short term, you can mitigate this by breaking the completion chain. Partway through the chain you can use `eventLoop.submit` inside your callback to prevent the callback returning a future that is immediately succeeded. That will force the stack to unwind.

**Comment by normanmaurer** [comment]
In netty we basically break the chain once we got a nesting of 8 by default. Maybe we should do something similar here.

The max “chain” size is configurable tho

**Comment by Lukasa** [comment]
Yup, that’s roughly the proposal I was thinking of.

**Comment by mxcl** [comment]
Thanks for the suggestion, but the submit trick doesn’t seem to work. I can conveniently do this often (I am parsing large numbers of `.jsonl` files), so I'd say I'm doing it every 8 or so callbacks.

I parse the line in `submit`, get vapor to save it in the database (so this is not an immediately succeeded future) and then return the function again to parse the next line.

**Comment by mxcl** [comment]
Here’s the psuedocode:

```swift
router.get("/insert", String.parameter) { rq in
    rq.transaction(on: .psql) { conn in
        try Iterator(state: try rq.parameters.next(String.self).uppercased()).next(on: rq)
    }
}

class Iterator {
    func next() -> EventLoopFuture<Void> {
        guard !done else { return conn.future() }
        return conn.eventLoop.submit {
            try self.decodeAModelLine()
        }.create(on: conn)
    }
}
```

**Comment by mxcl** [comment]
I figured out how to get Vapor to insert multiple records at once, so no longer need a workaround, thanks for the suggestion however.

**Comment by FredericRuaudel** [comment]
If it can helps anyone, here how I fixed my recursive method issue using `eventloop.submit` trick:

pseudocode:

```swift
func myRecursiveFunc() -> Future<Void> {
   return processFutureResult()
      .flatMap { done in
          if done { 
              return self.eventLoop.future() 
          } else {
              return self.eventLoop.submit { 
                   return myRecursiveFunc()
              }
        } 
}
```

The important part is that you need to wrap your call of the recursive method inside the `eventLoop.submit` closure.

Hope I didn't say too much nonsense 😉
